src/data/fin.lean

```diff
@@ -1012,6 +1012,168 @@ def append {α : Type*} {o : ℕ} (ho : o = m + n) (u : fin m → α) (v : fin n
   then u ⟨i, h⟩
   else v ⟨(i : ℕ) - m, (nat.sub_lt_left_iff_lt_add (le_of_not_lt h)).2 (ho ▸ i.property)⟩
 
+variables {ι : Type*} {o : ℕ} {u : fin m → ι} {v : fin n → ι}
+
+lemma append_apply_fst (ho : o = m + n) (k : fin o) (h : (k : ℕ) < m) :
+  append ho u v k = u (k.cast_lt h) :=
+dif_pos h
+
+@[simp] lemma append_apply_fst_cast (ho : o = m + n) (k : fin m) :
+  append ho u v (cast ho.symm (cast_add n k)) = u k :=
+(append_apply_fst ho (cast ho.symm (cast_add n k)) k.2).trans $
+  congr_arg _ (fin.ext rfl)
+
+lemma append_apply_snd (ho : o = m + n) (k : fin o) (h : ¬(k : ℕ) < m) :
+  append ho u v k = v ⟨(k : ℕ) - m, (nat.sub_lt_left_iff_lt_add $ le_of_not_lt h).2 (ho ▸ k.2)⟩ :=
+dif_neg h
+
+@[simp] lemma append_apply_snd_cast (ho : o = m + n) (k : fin n) :
+  append ho u v (cast ho.symm (nat_add m k)) = v k :=
+(append_apply_snd ho (cast ho.symm (nat_add m k)) (not_lt.2 $ le_add_right (le_refl _))).trans $
+  congr_arg _ (fin.ext $ nat.add_sub_cancel_left _ _)
+
+lemma append_default {d : fin 0 → ι} (ho : o = m) {x : fin o} :
+  append (show o = m + 0, from ho) u d x = u (cast ho x) :=
+(append_apply_fst (show o = m + 0, from ho) x (ho ▸ x.2)).trans $ by congr
```

Comment:
Is there precedent for this use of `default`? `fin 0 → _` seems to be referred to as `tuple0` above, so perhaps `append_tuple0` is a better choice of name.

Reply:
Also `show` should not be used in the statements of lemmas, because it inserts an identity function that can cause e.g. simp not to fire correctly. Use `(ho : o = m + 0)` instead

Reply:
Alternatively, #4406 uses the name `nil`

Reply:
> because it inserts an identity function that can cause e.g. simp not to fire correctly.

Is `simp` unable to match `hx` against `id hx` even though proof-irrelevance makes them equal?

Reply:
Precisely.  Lean needs to unify `proof_term_in_goal =?= id ?hx`, but it can only figure out what to assign to `?hx` after reducing `id`.  However `id` is not reducible.

That said, I'm not sure if `show ..., from hx` inserts an `id`.

For the lemma here, I think the best solution is to have `ho : o = m` as argument, and write `apply ho u d x`.  Or if that doesn't work, then `@apply ...`.